File: Python/Sorting/mergeSort.py
def merge(arr,start,mid,end):
    temp = [0]* (end-start+1)
    i,j,k = start,mid+1,0
    while i <=mid and j<=end:
        if(arr[i]<=arr[j]):
            temp[k]=arr[i]
            i+=1
        else:
            temp[k]=arr[j]
            j+=1
        k+=1
    while i <=mid:
        temp[k]=arr[i]
        i+=1
        k+=1
    while j<=end:
        temp[k]=arr[j]
        j+=1
        k+=1
    for i in range(start,end+1):
        arr[i]= temp[i-start]


def mergeSort(arr,start,end):
    if start < end:
        mid = int((start+end)/2)
        mergeSort(arr,start,mid)
        mergeSort(arr,mid+1,end)
        merge(arr,start,mid,end)

arr=[2,3,5,4,1]
mergeSort(arr,0,4)
print(arr)

File: C++/Sorting/MergeSort.cpp
/**
 * 
 * @description: Merge Sort O(nlogn)
 * 
 * @author: Lakshya Khera
 * 
 * */

#include<iostream>
#include<vector>
using namespace std;

void merge(vector<int> &arr, int l ,int m, int r) {
  int i, j, k;
  int n1 = m-l+1;
  int n2 = r-m;
  vector<int> L(n1),R(n2);
  for(i =0;i<n1;i++ ){
    L[i] = arr[l+i];    
  }
  for(j=0;j<n2;j++) {
    R[j] = arr[m+1+j];
  }
  i=0;
  j=0;
  k=l;
  while (i < n1 && j < n2) { 
    if (L[i] <= R[j]) { 
      arr[k] = L[i]; 
      i++; 
    } else { 
      arr[k] = R[j]; 
      j++; 
    } 
    k++; 
  } 
  while (i < n1) { 
    arr[k++] = L[i++]; 
  } 
  while (j < n2) { 
    arr[k++] = R[j++]; 
  } 

} 


void mergeSort(vector<int> &arr, int l ,int r) {
  if(l<r) {
    int mid = (l+r)/2;
    mergeSort(arr, l,mid);
    mergeSort(arr,mid+1,r);
    merge(arr,l,mid,r);
  }
}

int main() {

  vector<int> arr{12, 11, 13, 5, 6, 7 };
  mergeSort(arr,0,arr.size()-1);
  for(int i : arr)
    cout << i << " ";
  cout << endl;
  return 0;
}